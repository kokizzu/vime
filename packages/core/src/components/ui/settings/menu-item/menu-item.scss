vime-menu-item {
  /**
   * @prop --menu-item-hint-font-size: The font size of the hint text.
   * @prop --menu-item-hint-color: The color of the hint text.
   * @prop --menu-item-hint-opacity: The opacity of the hint text.
   * @prop --menu-item-badge-color: The color of the badge text.
   * @prop --menu-item-badge-bg: The background color of the badge.
   * @prop --menu-item-badge-font-size: The font size of the the badge text. 
   * @prop --menu-item-arrow-color: The color of the left/right arrow on a navigational menu item.
   * @prop --menu-item-check-icon-width: The width of the checked icon.
   * @prop --menu-item-check-icon-height: The height of the checked icon.
   * @prop --menu-item-padding: The padding within each menu item.
   * @prop --menu-item-focus-color: The color of a menu item's text when it is being focused or hovered on.
   * @prop --menu-item-focus-bg: The background color of a menu item when it is being focused or hovered on.
   * @prop --menu-item-tap-highlight: The highlight color when a menu item is tapped.
   * @prop --menu-item-divider-color: The color of the divider between a navigational menu item and the menu body.
   */

  display: flex;
  align-items: center;
  flex-direction: row;
  width: 100%;
  cursor: pointer;
  color: var(--menu-color);
  background: var(--menu-bg);
  font-size: var(--menu-font-size);
  font-weight: var(--menu-font-weight);
  padding: var(--menu-item-padding);

  &:focus {
    outline: 0;
  }

  &.hidden {
    display: none;
  }

  &.tapHighlight {
    background: var(--menu-item-tap-highlight);
  }

  &.showDivider {
    border-bottom: 0.5px solid var(--menu-item-divider-color);
  }

  &.notTouch {
    &:hover,
    &:focus {
      outline: 0;
      color: var(--menu-item-focus-color);
      background-color: var(--menu-item-focus-bg);
    }
  }

  &[aria-hidden='true'] {
    display: none;
  }

  vime-icon {
    display: inline-block;
  }

  svg {
    fill: currentColor;
    pointer-events: none;
    width: var(--menu-item-check-icon-width, 10px) !important;
    height: var(--menu-item-check-icon-height, 10px) !important;
    margin-right: 10px;
    @include invisible();
  }

  &[aria-checked='true'] svg {
    @include visible();
  }

  .hint {
    display: inline-block;
    margin-left: auto;
    overflow: hidden;
    pointer-events: none;
    margin-right: 6px;
    font-size: var(--menu-item-hint-font-size);
    opacity: var(--menu-item-hint-opacity);
    color: var(--menu-item-hint-color);
  }
  
  .badge {
    display: inline-block;
    line-height: 1;
    overflow: hidden;
    pointer-events: none;
    color: var(--menu-item-badge-color);
    background: var(--menu-item-badge-bg);
    font-size: var(--menu-item-badge-font-size);
  }

  .spacer {
    flex: 1;
  }

  .arrow {
    color: var(--menu-item-arrow-color);
    border: 2px solid;
    padding: 2px;
    display: inline-block;
    border-width: 0 2px 2px 0;
  
    &.left {
      margin-right: 6px;
      transform: rotate(135deg);
    }
  
    &.right {
      transform: rotate(-45deg);
      opacity: 0.38;
    }
  }
}
