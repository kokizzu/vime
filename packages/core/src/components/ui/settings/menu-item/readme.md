# vime-menu-item

A multi-purpose interactable element inside a menu. The behaviour and style of the item depends on
the props set.

- **Default:** By default, the menu item only contains a label and optional hint/badge text that is
  displayed on the right-hand side of the item.
- **Navigation:** If the `menu` prop is set, the item behaves as a navigational control and displays
  arrows to indicate whether clicking the control will navigate forwards/backwards.
- **Radio:** If the `checked` prop is set, the item behaves as a radio button and displays a
  checkmark icon to indicate whether it is checked or not.

## Visual

<img
  src="https://raw.githubusercontent.com/vime-js/vime/master/packages/core/src/components/ui/settings/menu-item/menu-item.png"
  alt="Vime settings menu item component"
/>

<!-- Auto Generated Below -->

## Usage

### Angular

```html {6} title="example.html"
<vime-player>
  <!-- ... -->
  <vime-ui>
    <!-- ... -->
    <vime-settings>
      <vime-menu-item label="Playback Quality" hint="Auto"></vime-menu-item>
    </vime-settings>
  </vime-ui>
</vime-player>
```

### Html

```html {6}
<vime-player>
  <!-- ... -->
  <vime-ui>
    <!-- ... -->
    <vime-settings>
      <vime-menu-item label="Playback Quality" hint="Auto"></vime-menu-item>
    </vime-settings>
  </vime-ui>
</vime-player>
```

### React

```tsx {2,11}
import React from 'react';
import { VimePlayer, VimeUi, VimeSettings, VimeMenuItem } from '@vime/react';

function Example() {
  return render(
    <VimePlayer>
      {/* ... */}
      <VimeUi>
        {/* ... */}
        <VimeSettings>
          <VimeMenuItem label="Playback Quality" hint="Auto" />
        </VimeSettings>
      </VimeUi>
    </VimePlayer>
  );
}
```

### Vue

```html {7,14,21} title="example.vue"
<template>
  <VimePlayer>
    <!-- ... -->
    <VimeUi>
      <!-- ... -->
      <VimeSettings>
        <VimeMenuItem label="Playback Quality" hint="Auto" />
      </VimeSettings>
    </VimeUi>
  </VimePlayer>
</template>

<script>
  import { VimePlayer, VimeUi, VimeSettings, VimeMenuItem } from '@vime/vue';

  export default {
    components: {
      VimePlayer,
      VimeUi,
      VimeSettings,
      VimeMenuItem,
    },
  };
</script>
```

## Properties

| Property             | Attribute      | Description                                                                                                                                                                                               | Type                   | Default             |
| -------------------- | -------------- | --------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- | ---------------------- | ------------------- |
| `badge`              | `badge`        | This can provide additional context about the value of a menu item. For example, if the item is a radio button for a set of video qualities, the badge could describe whether the quality is UHD, HD etc. | `string \| undefined`  | `undefined`         |
| `checked`            | `checked`      | If this item is to behave as a radio button, then this property determines whether the radio is selected or not. Sets the `aria-checked` property.                                                        | `boolean \| undefined` | `undefined`         |
| `checkedIcon`        | `checked-icon` | The URL to an SVG element or fragment to load.                                                                                                                                                            | `string \| undefined`  | `'#vime-checkmark'` |
| `expanded`           | `expanded`     | If the item has a popup menu, this indicates whether the menu is open or not. Sets the `aria-expanded` property.                                                                                          | `boolean \| undefined` | `undefined`         |
| `hidden`             | `hidden`       | Whether the item is displayed or not.                                                                                                                                                                     | `boolean`              | `false`             |
| `hint`               | `hint`         | This can provide additional context about some underlying state of the item. For example, if the menu item opens/closes a submenu with options, the hint could be the currently selected option.          | `string \| undefined`  | `undefined`         |
| `identifier`         | `identifier`   | The `id` attribute of the item.                                                                                                                                                                           | `string \| undefined`  | `undefined`         |
| `label` _(required)_ | `label`        | The label/title of the item.                                                                                                                                                                              | `string`               | `undefined`         |
| `menu`               | `menu`         | If the item has a popup menu, then this should be the `id` of said menu. Sets the `aria-controls` property.                                                                                               | `string \| undefined`  | `undefined`         |

## CSS Custom Properties

| Name                            | Description                                                                  |
| ------------------------------- | ---------------------------------------------------------------------------- |
| `--menu-item-arrow-color`       | The color of the left/right arrow on a navigational menu item.               |
| `--menu-item-badge-bg`          | The background color of the badge.                                           |
| `--menu-item-badge-color`       | The color of the badge text.                                                 |
| `--menu-item-badge-font-size`   | The font size of the the badge text.                                         |
| `--menu-item-check-icon-height` | The height of the checked icon.                                              |
| `--menu-item-check-icon-width`  | The width of the checked icon.                                               |
| `--menu-item-divider-color`     | The color of the divider between a navigational menu item and the menu body. |
| `--menu-item-focus-bg`          | The background color of a menu item when it is being focused or hovered on.  |
| `--menu-item-focus-color`       | The color of a menu item's text when it is being focused or hovered on.      |
| `--menu-item-hint-color`        | The color of the hint text.                                                  |
| `--menu-item-hint-font-size`    | The font size of the hint text.                                              |
| `--menu-item-hint-opacity`      | The opacity of the hint text.                                                |
| `--menu-item-padding`           | The padding within each menu item.                                           |
| `--menu-item-tap-highlight`     | The highlight color when a menu item is tapped.                              |

## Dependencies

### Used by

- [vime-default-settings](../default-settings)
- [vime-menu-radio](../menu-radio)
- [vime-submenu](../submenu)

### Depends on

- [vime-icon](../../icon)

### Graph

```mermaid
graph TD;
  vime-menu-item --> vime-icon
  vime-default-settings --> vime-menu-item
  vime-menu-radio --> vime-menu-item
  vime-submenu --> vime-menu-item
  style vime-menu-item fill:#f9f,stroke:#333,stroke-width:4px
```

---

_Built with [StencilJS](https://stenciljs.com/)_
