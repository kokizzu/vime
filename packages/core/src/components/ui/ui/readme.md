# vime-ui

Simple container that holds a collection of user interface components.

The only important role this component really has is, avoiding overlapping custom UI with the 
native iOS media player UI. Therefore, custom UI is only displayed on iOS if the `playsinline` prop 
is `true`, and the player is not in fullscreen mode.

## Example

```html
<vime-player>
  <!-- ... -->

  <vime-ui>
    <vime-poster />
    <!-- ... -->
  </vime-ui>
</vime-player>
```

<!-- Auto Generated Below -->


## Dependencies

### Used by

 - [vime-default-ui](../default-ui)

### Graph
```mermaid
graph TD;
  vime-default-ui --> vime-ui
  style vime-ui fill:#f9f,stroke:#333,stroke-width:4px
```

----------------------------------------------

*Built with [StencilJS](https://stenciljs.com/)*
