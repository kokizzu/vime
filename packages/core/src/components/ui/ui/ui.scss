vime-ui {
  position: relative;
  pointer-events: none;
  z-index: $z-index-ui;
  width: 100%;

  &.hidden {
    display: none;
  }

  &.video {
    position: absolute;
    top: 0;
    left: 0;
    height: 100%;
  }
}