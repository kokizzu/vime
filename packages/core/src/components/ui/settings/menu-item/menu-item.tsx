/* eslint-disable jsx-a11y/no-noninteractive-tabindex */
/* eslint-disable jsx-a11y/role-supports-aria-props */

import {
  h, Host, Component, Prop, State, Element,
} from '@stencil/core';
import { isUndefined, isNull } from '../../../../utils/unit';
import { PlayerProps, PlayerProp } from '../../../core/player/PlayerProp';
import { openPlayerWormhole } from '../../../core/player/PlayerWormhole';

@Component({
  tag: 'vime-menu-item',
  styleUrl: 'menu-item.scss',
})
export class MenuItem {
  @Element() el!: HTMLVimeMenuItemElement;

  @State() showTapHighlight = false;

  /**
   * The `id` attribute of the item.
   */
  @Prop() identifier?: string;

  /**
   * Whether the item is displayed or not.
   */
  @Prop() hidden = false;

  /**
   * The label/title of the item.
   */
  @Prop() label!: string;

  /**
   * If the item has a popup menu, then this should be the `id` of said menu. Sets the
   * `aria-controls` property.
   */
  @Prop() menu?: string;

  /**
   * If the item has a popup menu, this indicates whether the menu is open or not. Sets the
   * `aria-expanded` property.
   */
  @Prop() expanded?: boolean;

  /**
   * If this item is to behave as a radio button, then this property determines whether the
   * radio is selected or not. Sets the `aria-checked` property.
   */
  @Prop() checked?: boolean;

  /**
   * This can provide additional context about some underlying state of the item. For example, if
   * the menu item opens/closes a submenu with options, the hint could be the currently selected
   * option.
   */
  @Prop() hint?: string;

  /**
   * This can provide additional context about the value of a menu item. For example, if the item
   * is a radio button for a set of video qualities, the badge could describe whether the quality
   * is UHD, HD etc.
   */
  @Prop() badge?: string;

  /**
   * The URL to an SVG element or fragment to load.
   */
  @Prop() checkedIcon?: string = '#vime-checkmark';

  /**
   * @internal
   */
  @Prop() isTouch: PlayerProps[PlayerProp.IsTouch] = false;

  private onClick() {
    if (isUndefined(this.menu)) return;
    const submenu = document.querySelector(`#${this.menu}`) as HTMLVimeMenuElement;
    if (!isNull(submenu)) submenu!.active = !this.expanded;
  }

  private onTouchStart() {
    this.showTapHighlight = true;
    setTimeout(() => { this.showTapHighlight = false; }, 100);
  }

  private onMouseLeave() {
    this.el.blur();
  }

  render() {
    const isCheckedDefined = !isUndefined(this.checked);
    const isMenuDefined = !isUndefined(this.menu);
    const hasExpanded = this.expanded ? 'true' : 'false';
    const isChecked = this.checked ? 'true' : 'false';
    const showCheckedIcon = isCheckedDefined && !isUndefined(this.checkedIcon);
    const showLeftNavArrow = isMenuDefined && this.expanded;
    const showRightNavArrow = isMenuDefined && !this.expanded;
    const showHint = !isUndefined(this.hint)
      && !isCheckedDefined
      && (!isMenuDefined || !this.expanded);
    const showBadge = !isUndefined(this.badge) && (!showHint && !showRightNavArrow);
    const hasSpacer = showHint || showBadge || showRightNavArrow;

    return (
      <Host
        class={{
          notTouch: !this.isTouch,
          tapHighlight: this.showTapHighlight,
          showDivider: isMenuDefined && (this.expanded ?? false),
        }}
        id={this.identifier}
        role={isCheckedDefined ? 'menuitemradio' : 'menuitem'}
        tabindex="0"
        aria-label={this.label}
        aria-hidden={this.hidden ? 'true' : 'false'}
        aria-haspopup={isMenuDefined ? 'true' : undefined}
        aria-controls={this.menu}
        aria-expanded={isMenuDefined ? hasExpanded : undefined}
        aria-checked={isCheckedDefined ? isChecked : undefined}
        onClick={this.onClick.bind(this)}
        onTouchStart={this.onTouchStart.bind(this)}
        onMouseLeave={this.onMouseLeave.bind(this)}
      >
        {showCheckedIcon && <vime-icon href={this.checkedIcon!} />}
        {showLeftNavArrow && <span class="arrow left" />}
        {this.label}
        {hasSpacer && <span class="spacer" />}
        {showHint && <span class="hint">{this.hint}</span>}
        {showBadge && <span class="badge">{this.badge}</span>}
        {showRightNavArrow && <span class="arrow right" />}
      </Host>
    );
  }
}

openPlayerWormhole(MenuItem, [
  PlayerProp.IsTouch,
]);
