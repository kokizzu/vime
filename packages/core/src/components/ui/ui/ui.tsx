import {
  h, Host, Component, Prop,
} from '@stencil/core';
import { withPlayerContext } from '../../core/player/PlayerContext';
import { PlayerProp, PlayerProps } from '../../core/player/PlayerProp';
import { IS_IOS } from '../../../utils/support';

/**
 * @slot - Used to pass in UI components for the player.
 */
@Component({
  tag: 'vime-ui',
  styleUrl: 'ui.scss',
})
export class UI {
  /**
   * @internal
   */
  @Prop() isVideoView: PlayerProps[PlayerProp.isVideoView] = false;

  /**
   * @internal
   */
  @Prop() playsinline: PlayerProps[PlayerProp.playsinline] = false;

  /**
   * @internal
   */
  @Prop() isFullscreenActive: PlayerProps[PlayerProp.isFullscreenActive] = false;

  render() {
    const canShowCustomUI = !IS_IOS
      || !this.isVideoView
      || (this.playsinline && !this.isFullscreenActive);

    return (
      <Host
        class={{
          hidden: !canShowCustomUI,
          video: this.isVideoView,
        }}
      >
        <div>
          {canShowCustomUI && <slot />}
        </div>
      </Host>
    );
  }
}

withPlayerContext(UI, [
  PlayerProp.isVideoView,
  PlayerProp.playsinline,
  PlayerProp.isFullscreenActive,
]);
