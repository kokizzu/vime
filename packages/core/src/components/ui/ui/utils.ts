import { getElement } from '@stencil/core';
import { HTMLStencilElement } from '@stencil/core/internal';

export const findUIRoot = (ref: any) => {
  let ui = getElement(ref);

  while (!(/^VIME-UI$/.test(ui?.nodeName))) {
    ui = ui.parentElement as HTMLStencilElement;
  }

  return ui as HTMLVimeUiElement;
};
