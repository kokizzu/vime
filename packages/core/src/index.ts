import { PlayerProp, PlayerProps } from './components/core/player/PlayerProp';
import { PlayerDispatcher, createPlayerDispatcher } from './components/core/player/PlayerDispatcher';
import { withPlayerContext, usePlayerContext } from './components/core/player/PlayerContext';
import { PlayerEvent } from './components/core/player/PlayerEvent';
import { findRootPlayer } from './components/core/player/utils';

export * from './components';

export {
  PlayerProp,
  PlayerProps,
  PlayerEvent,
  findRootPlayer,
  withPlayerContext,
  usePlayerContext,
  PlayerDispatcher,
  createPlayerDispatcher,
};
