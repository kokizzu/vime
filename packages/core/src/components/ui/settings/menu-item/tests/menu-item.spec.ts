import { SpecPage } from '@stencil/core/testing';
import { MenuItem } from '../menu-item';
import { newUISpecPage } from '../../../ui/tests';

let page: SpecPage;
let menuItem: HTMLVimeMenuItemElement;

beforeEach(async () => {
  ({ page } = await newUISpecPage(
    [MenuItem],
    '<vime-menu-item label="Label"></vime-menu-item>',
  ));

  menuItem = page.root!.querySelector('vime-menu-item')! as HTMLVimeMenuItemElement;
});

it('should be structurally sound if plain', () => {
  expect(menuItem).toMatchSnapshot();
});

it('should be structurally sound if navigation item', async () => {
  menuItem.menu = 'menu-1';
  menuItem.hint = 'hint';
  await page.waitForChanges();
  expect(menuItem).toMatchSnapshot();
});

it('should be structurally sound if expanded navigation item', async () => {
  menuItem.menu = 'menu-1';
  menuItem.expanded = true;
  menuItem.hint = 'hint';
  await page.waitForChanges();
  expect(menuItem).toMatchSnapshot();
});

it('should be structurally sound if checked radio button', async () => {
  menuItem.checked = true;
  await page.waitForChanges();
  expect(menuItem).toMatchSnapshot();
});

it('should be structurally sound if unchecked radio button', async () => {
  menuItem.checked = false;
  await page.waitForChanges();
  expect(menuItem).toMatchSnapshot();
});

it('should be structurally sound if item has hint', async () => {
  menuItem.hint = 'hint';
  await page.waitForChanges();
  expect(menuItem).toMatchSnapshot();
});

it('should be structurally sound if item has a badge', async () => {
  menuItem.badge = 'badge';
  await page.waitForChanges();
  expect(menuItem).toMatchSnapshot();
});
