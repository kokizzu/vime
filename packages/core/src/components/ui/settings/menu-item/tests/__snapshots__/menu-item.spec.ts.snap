// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should be structurally sound if checked radio button 1`] = `
<vime-menu-item aria-checked="true" aria-haspopup="false" aria-hidden="false" aria-label="Label" class="notTouch" label="Label" role="menuitemradio" tabindex="0">
  <vime-icon href="#vime-checkmark"></vime-icon>
  Label
</vime-menu-item>
`;

exports[`should be structurally sound if expanded navigation item 1`] = `
<vime-menu-item aria-controls="menu-1" aria-expanded="true" aria-haspopup="true" aria-hidden="false" aria-label="Label" class="notTouch showDivider" label="Label" role="menuitem" tabindex="0">
  <span class="arrow left"></span>
  Label
</vime-menu-item>
`;

exports[`should be structurally sound if item has a badge 1`] = `
<vime-menu-item aria-haspopup="false" aria-hidden="false" aria-label="Label" class="notTouch" label="Label" role="menuitem" tabindex="0">
  Label
  <span class="spacer"></span>
  <span class="badge">
    badge
  </span>
</vime-menu-item>
`;

exports[`should be structurally sound if item has hint 1`] = `
<vime-menu-item aria-haspopup="false" aria-hidden="false" aria-label="Label" class="notTouch" label="Label" role="menuitem" tabindex="0">
  Label
  <span class="spacer"></span>
  <span class="hint">
    hint
  </span>
</vime-menu-item>
`;

exports[`should be structurally sound if navigation item 1`] = `
<vime-menu-item aria-controls="menu-1" aria-expanded="false" aria-haspopup="true" aria-hidden="false" aria-label="Label" class="notTouch" label="Label" role="menuitem" tabindex="0">
  Label
  <span class="spacer"></span>
  <span class="hint">
    hint
  </span>
  <span class="arrow right"></span>
</vime-menu-item>
`;

exports[`should be structurally sound if plain 1`] = `
<vime-menu-item aria-haspopup="false" aria-hidden="false" aria-label="Label" class="notTouch" label="Label" role="menuitem" tabindex="0">
  Label
</vime-menu-item>
`;

exports[`should be structurally sound if unchecked radio button 1`] = `
<vime-menu-item aria-checked="false" aria-haspopup="false" aria-hidden="false" aria-label="Label" class="notTouch" label="Label" role="menuitemradio" tabindex="0">
  <vime-icon href="#vime-checkmark"></vime-icon>
  Label
</vime-menu-item>
`;
